crates/typst/src/text/font/color.rs

```diff
@@ -0,0 +1,321 @@
+//! Utilities for color font handling
+
+use std::io::Read;
+
+use ecow::EcoString;
+use ttf_parser::GlyphId;
+use usvg::{Tree, TreeParsing, TreeWriting};
+
+use crate::layout::{Abs, Axes, Em, Frame, FrameItem, Point, Size};
+use crate::syntax::Span;
+use crate::text::{Glyph, Lang, TextItem, TextItemView};
+use crate::visualize::{Color, Image, Paint, Rgb};
+
+use super::Font;
+
+/// Tells if a glyph is a color glyph or not in a given font.
+pub fn is_color_glyph(font: &Font, g: &Glyph) -> bool {
+    let ttf = font.ttf();
+    let glyph_id = GlyphId(g.id);
+    ttf.glyph_raster_image(glyph_id, 160).is_some()
+        || ttf.glyph_svg_image(glyph_id).is_some()
+        || ttf.is_color_glyph(glyph_id)
+}
+
+/// Returns a frame with the glyph drawn inside.
+///
+/// The glyphs are sized in font units, [`text.item.size`] is not taken into
+/// account.
+pub fn frame_for_glyph(text: &TextItemView, glyph: &Glyph) -> Frame {
+    #[comemo::memoize]
+    fn cached(font: &Font, fill: &Paint, glyph_id: u16) -> Frame {
+        let ttf = font.ttf();
+        let upem = Abs::pt(ttf.units_per_em() as f64);
+        let glyph_id = GlyphId(glyph_id);
+
+        let mut frame = Frame::soft(Size::splat(upem));
+
+        if let Some(raster_image) = ttf.glyph_raster_image(glyph_id, u16::MAX) {
+            draw_raster_glyph(&mut frame, upem, raster_image);
+        } else if ttf.glyph_svg_image(glyph_id).is_some() {
+            draw_svg_glyph(&mut frame, upem, font, glyph_id);
+        } else if ttf.is_color_glyph(glyph_id) {
+            draw_colr_glyph(&mut frame, font, fill, glyph_id);
+        }
+
+        frame
+    }
+
+    cached(&text.item.font, &text.item.fill, glyph.id)
+}
+
+/// Draws a raster glyph in a frame.
+fn draw_raster_glyph(
+    frame: &mut Frame,
+    upem: Abs,
+    raster_image: ttf_parser::RasterGlyphImage,
+) {
+    let image = Image::new(
+        raster_image.data.into(),
+        typst::visualize::ImageFormat::Raster(typst::visualize::RasterFormat::Png),
+        None,
+    )
+    .unwrap();
+    let position = Point::new(
+        upem * raster_image.x as f64 / raster_image.pixels_per_em as f64,
+        upem * -raster_image.y as f64 / raster_image.pixels_per_em as f64,
```

Comment:
might be a little safer to negate on the float side than on i16 because of potential for overflow, though it's probably not a thing that'll happen in practice